[Bug|Question] Conflict caused by FALSE, TRUE macros and bool_tag enum using same name

I need to build Universal Ctags on an AIX 7.2 system using gcc version 8.3.0. During the **make** step I am getting the error

```
In file included from /opt/freeware/lib/gcc/powerpc-ibm-aix7.2.0.0/8.3.0/include-fixed/stdio.h:503,
               from ./misc/packcc/src/packcc.c:40:
./misc/packcc/src/packcc.c:94:5: error: expected identifier before numeric constant
   FALSE = 0,
   ^~~~~
```
I have looked into the pre-processed file packcc.i. The indentifier FALSE was replaced with 0. Conclusion: Some header file defines the macros FALSE and TRUE that create a conflict with the enum definition

94 typedef enum bool_tag {
95     FALSE = 0,
96     TRUE
97 } bool_t;

inside packcc.c. I was able to overcome the issue #undef ing the macros just before the enum definition. This obviously isn't a desirable thing to do.

What is the correct course of action for me to take?

--- a/Makefile.am
+++ b/Makefile.am
@@ -48,8 +48,9 @@ noinst_LIBRARIES = libctags.a
 noinst_PROGRAMS =
 noinst_PROGRAMS += mini-geany
 
+PACKCC_CPPFLAGS_FOR_BUILD  = -fsigned-char
 if HAVE_STRNLEN_FOR_BUILD
-EXTRA_CPPFLAGS_FOR_BUILD = -DPCC_USE_SYSTEM_STRNLEN
+PACKCC_CPPFLAGS_FOR_BUILD += -DPCC_USE_SYSTEM_STRNLEN
 endif
 
 PACKCC = $(top_builddir)/packcc$(BUILD_EXEEXT)
@@ -58,7 +59,7 @@ cc4b_verbose = $(cc4b_verbose_@AM_V@)
 cc4b_verbose_ = $(cc4b_verbose_@AM_DEFAULT_V@)
 cc4b_verbose_0 = @echo CC4BUILD "  $@";
 $(PACKCC): $(top_srcdir)/misc/packcc/src/packcc.c
-	$(cc4b_verbose)$(CC_FOR_BUILD) $(CFLAGS_FOR_BUILD) $(CPPFLAGS_FOR_BUILD) $(EXTRA_CPPFLAGS_FOR_BUILD) $(LDFLAGS_FOR_BUILD) -o $@ $(top_srcdir)/misc/packcc/src/packcc.c
+	$(cc4b_verbose)$(CC_FOR_BUILD) $(CFLAGS_FOR_BUILD) $(CPPFLAGS_FOR_BUILD) $(PACKCC_CPPFLAGS_FOR_BUILD) $(LDFLAGS_FOR_BUILD) -o $@ $(top_srcdir)/misc/packcc/src/packcc.c
 
 if USE_READCMD
 bin_PROGRAMS+= readtags

--- a/misc/packcc/README.md
+++ b/misc/packcc/README.md
@@ -2,7 +2,7 @@
 
 ## Overview ##
 
-**PackCC** is a packrat parser generator for C. Its main features are as follows:
+**PackCC** is a parser generator for C. Its main features are as follows:
 
 - Generates your parser in C from a grammar described in a **PEG**,
 - Gives your parser great efficiency by **packrat parsing**,
@@ -34,6 +34,8 @@ For convenience, the build environments using GCC, Clang, and Microsoft Visual S
 
 ### Using GCC ###
 
+#### Other than MinGW ####
+
 `packcc` will be built in both directories `build/gcc/debug/bin` and `build/gcc/release/bin` using `gcc` by executing the following commands:
 
 ```
@@ -44,7 +46,21 @@ make check  # bats-core and uncrustify are required (see tests/README.md)
 
 `packcc` in the directory `build/gcc/release/bin` is suitable for practical use.
 
+#### MinGW ####
+
+`packcc` will be built in both directories `build/mingw-gcc/debug/bin` and `build/mingw-gcc/release/bin` using `gcc` by executing the following commands:
+
+```
+cd build/mingw-gcc
+make
+make check  # bats-core and uncrustify are required (see tests/README.md)
+```
+
+`packcc` in the directory `build/mingw-gcc/release/bin` is suitable for practical use.
+
 ### Using Clang ###
+
+#### Other than MinGW ####
 
 `packcc` will be built in both directories `build/clang/debug/bin` and `build/clang/release/bin` using `clang` by executing the following commands:
 
@@ -55,6 +71,18 @@ make check  # bats-core and uncrustify are required (see tests/README.md)
 ```
 
 `packcc` in the directory `build/clang/release/bin` is suitable for practical use.
+
+#### MinGW ####
+
+`packcc` will be built in both directories `build/mingw-clang/debug/bin` and `build/mingw-clang/release/bin` using `clang` by executing the following commands:
+
+```
+cd build/mingw-clang
+make
+make check  # bats-core and uncrustify are required (see tests/README.md)
+```
+
+`packcc` in the directory `build/mingw-clang/release/bin` is suitable for practical use.
 
 ### Using Microsoft Visual Studio ###
 

--- a/misc/packcc/build/clang/Makefile
+++ b/misc/packcc/build/clang/Makefile
@@ -1,6 +1,6 @@
 CC=clang
-CFLAGS_D=-std=gnu89 -Wall -Wextra -Wno-unused-parameter -Wno-overlength-strings -pedantic -O0 -g2
-CFLAGS_R=-std=gnu89 -Wall -Wextra -Wno-unused-parameter -Wno-overlength-strings -pedantic -O2 -DNDEBUG
+CFLAGS_D=-std=gnu89 -fsigned-char -Wall -Wextra -Wno-unused-parameter -Wno-overlength-strings -pedantic -O0 -g2
+CFLAGS_R=-std=gnu89 -fsigned-char -Wall -Wextra -Wno-unused-parameter -Wno-overlength-strings -pedantic -O2 -DNDEBUG
 LDFLAGS_D=
 LDFLAGS_R=
 
@@ -24,7 +24,7 @@ SRCS= \
   $(patsubst %,$(TMPDIR_R)/%.c,$(EXAMPLES)) \
   $(patsubst %,$(TMPDIR_R)/%.h,$(EXAMPLES))
 
-.PHONY: all clean
+.PHONY: all check clean
 
 .SECONDARY: $(SRCS)
 
@@ -48,8 +48,9 @@ $(TMPDIR_D)/examples/%.c $(TMPDIR_D)/examples/%.h: $(SRCDIR)/examples/%.peg $(BI
 $(TMPDIR_R)/examples/%.c $(TMPDIR_R)/examples/%.h: $(SRCDIR)/examples/%.peg $(BINDIR_R)/packcc
 	mkdir -p $(dir $@) && $(BINDIR_R)/packcc -o $(basename $@) $<
 
-check: $(BINDIR_R)/packcc
-	PACKCC=$$PWD/$^ $(TESTDIR)/test.sh
+check: $(BINDIR_D)/packcc $(BINDIR_R)/packcc
+	@echo "== Debug Version =="   && PACKCC=$$PWD/$(BINDIR_D)/packcc CC="$(CC) $(CFLAGS_D)" $(TESTDIR)/test.sh
+	@echo "== Release Version ==" && PACKCC=$$PWD/$(BINDIR_R)/packcc CC="$(CC) $(CFLAGS_R)" $(TESTDIR)/test.sh
 
 clean:
 	rm -f $(BINS) $(SRCS)

--- a/misc/packcc/build/gcc/Makefile
+++ b/misc/packcc/build/gcc/Makefile
@@ -1,6 +1,6 @@
 CC=gcc
-CFLAGS_D=-std=gnu89 -Wall -Wextra -Wno-unused-parameter -Wno-overlength-strings -pedantic -O0 -g2
-CFLAGS_R=-std=gnu89 -Wall -Wextra -Wno-unused-parameter -Wno-overlength-strings -pedantic -O2 -DNDEBUG
+CFLAGS_D=-std=gnu89 -fsigned-char -Wall -Wextra -Wno-unused-parameter -Wno-overlength-strings -pedantic -O0 -g2
+CFLAGS_R=-std=gnu89 -fsigned-char -Wall -Wextra -Wno-unused-parameter -Wno-overlength-strings -pedantic -O2 -DNDEBUG
 LDFLAGS_D=
 LDFLAGS_R=
 
@@ -24,7 +24,7 @@ SRCS= \
   $(patsubst %,$(TMPDIR_R)/%.c,$(EXAMPLES)) \
   $(patsubst %,$(TMPDIR_R)/%.h,$(EXAMPLES))
 
-.PHONY: all clean
+.PHONY: all check clean
 
 .SECONDARY: $(SRCS)
 
@@ -48,8 +48,9 @@ $(TMPDIR_D)/examples/%.c $(TMPDIR_D)/examples/%.h: $(SRCDIR)/examples/%.peg $(BI
 $(TMPDIR_R)/examples/%.c $(TMPDIR_R)/examples/%.h: $(SRCDIR)/examples/%.peg $(BINDIR_R)/packcc
 	mkdir -p $(dir $@) && $(BINDIR_R)/packcc -o $(basename $@) $<
 
-check: $(BINDIR_R)/packcc
-	PACKCC=$$PWD/$^ $(TESTDIR)/test.sh
+check: $(BINDIR_D)/packcc $(BINDIR_R)/packcc
+	@echo "== Debug Version =="   && PACKCC=$$PWD/$(BINDIR_D)/packcc CC="$(CC) $(CFLAGS_D)" $(TESTDIR)/test.sh
+	@echo "== Release Version ==" && PACKCC=$$PWD/$(BINDIR_R)/packcc CC="$(CC) $(CFLAGS_R)" $(TESTDIR)/test.sh
 
 clean:
 	rm -f $(BINS) $(SRCS)

--- a/misc/packcc/build/mingw-clang/Makefile
+++ b/misc/packcc/build/mingw-clang/Makefile
@@ -24,7 +24,7 @@ SRCS= \
   $(patsubst %,$(TMPDIR_R)/%.c,$(EXAMPLES)) \
   $(patsubst %,$(TMPDIR_R)/%.h,$(EXAMPLES))
 
-.PHONY: all clean
+.PHONY: all check clean
 
 .SECONDARY: $(SRCS)
 
@@ -48,8 +48,9 @@ $(TMPDIR_D)/examples/%.c $(TMPDIR_D)/examples/%.h: $(SRCDIR)/examples/%.peg $(BI
 $(TMPDIR_R)/examples/%.c $(TMPDIR_R)/examples/%.h: $(SRCDIR)/examples/%.peg $(BINDIR_R)/packcc
 	mkdir -p $(dir $@) && $(BINDIR_R)/packcc -o $(basename $@) $<
 
-check: $(BINDIR_R)/packcc
-	PACKCC=$$PWD/$^ $(TESTDIR)/test.sh
+check: $(BINDIR_D)/packcc $(BINDIR_R)/packcc
+	@echo "== Debug Version =="   && PACKCC=$$PWD/$(BINDIR_D)/packcc CC="$(CC) $(CFLAGS_D)" $(TESTDIR)/test.sh
+	@echo "== Release Version ==" && PACKCC=$$PWD/$(BINDIR_R)/packcc CC="$(CC) $(CFLAGS_R)" $(TESTDIR)/test.sh
 
 clean:
 	rm -f $(BINS) $(SRCS)

--- a/misc/packcc/build/mingw-gcc/Makefile
+++ b/misc/packcc/build/mingw-gcc/Makefile
@@ -24,7 +24,7 @@ SRCS= \
   $(patsubst %,$(TMPDIR_R)/%.c,$(EXAMPLES)) \
   $(patsubst %,$(TMPDIR_R)/%.h,$(EXAMPLES))
 
-.PHONY: all clean
+.PHONY: all check clean
 
 .SECONDARY: $(SRCS)
 
@@ -48,8 +48,9 @@ $(TMPDIR_D)/examples/%.c $(TMPDIR_D)/examples/%.h: $(SRCDIR)/examples/%.peg $(BI
 $(TMPDIR_R)/examples/%.c $(TMPDIR_R)/examples/%.h: $(SRCDIR)/examples/%.peg $(BINDIR_R)/packcc
 	mkdir -p $(dir $@) && $(BINDIR_R)/packcc -o $(basename $@) $<
 
-check: $(BINDIR_R)/packcc
-	PACKCC=$$PWD/$^ $(TESTDIR)/test.sh
+check: $(BINDIR_D)/packcc $(BINDIR_R)/packcc
+	@echo "== Debug Version =="   && PACKCC=$$PWD/$(BINDIR_D)/packcc CC="$(CC) $(CFLAGS_D)" $(TESTDIR)/test.sh
+	@echo "== Release Version ==" && PACKCC=$$PWD/$(BINDIR_R)/packcc CC="$(CC) $(CFLAGS_R)" $(TESTDIR)/test.sh
 
 clean:
 	rm -f $(BINS) $(SRCS)

--- a/misc/packcc/src/packcc.c
+++ b/misc/packcc/src/packcc.c
@@ -68,6 +68,9 @@ static size_t strnlen_(const char *str, size_t maxlen) {
 #define __attribute__(x)
 #endif
 
+#undef TRUE  /* to avoid macro definition conflicts with the system header file of IBM AIX */
+#undef FALSE
+
 #define VERSION "1.5.0"
 
 #ifndef BUFFER_INIT_SIZE
@@ -910,7 +913,7 @@ static size_t populate_bits(size_t x) {
     x |= x >>  4;
     x |= x >>  8;
     x |= x >> 16;
-#ifndef _M_IX86 /* not Windows for x86 (32-bit) */
+#if (defined __SIZEOF_SIZE_T__ && __SIZEOF_SIZE_T__ == 8) /* gcc or clang */ || defined _WIN64 /* MSVC */
     x |= x >> 32;
 #endif
     return x;

--- a/misc/packcc/tests/.gitignore
+++ b/misc/packcc/tests/.gitignore
@@ -1,6 +1,8 @@
 packcc
 packcc.exe
-*/parser.peg
+packcc.gcda
+packcc.gcno
+packcc.c.gcov
 */parser.c
 */parser.h
 */parser

--- a/misc/packcc/tests/ascii.d/ascii.bats
+++ b/misc/packcc/tests/ascii.d/ascii.bats
@@ -1,0 +1,19 @@
+#!/usr/bin/env bats
+
+load "$TESTDIR/utils.sh"
+
+@test "Testing ascii.d - generation" {
+    PACKCC_OPTS=("--ascii")
+    test_generate
+}
+
+@test "Testing ascii.d - check code" {
+    ! in_source "pcc_get_char_as_utf32"
+}
+
+@test "Testing ascii.d - compilation" {
+    test_compile
+}
+@test "Testing ascii.d - run" {
+    run_for_input "ascii.d/input.txt"
+}

--- a/misc/packcc/tests/ascii.d/expected.txt
+++ b/misc/packcc/tests/ascii.d/expected.txt
@@ -1,0 +1,4 @@
+This
+is
+a
+test

--- a/misc/packcc/tests/ascii.d/input.peg
+++ b/misc/packcc/tests/ascii.d/input.peg
@@ -1,0 +1,3 @@
+FILE <- WORD / SPACE
+WORD <- [^ \r\n\t]+ { PRINT($0); }
+SPACE <- [ \r\n\t]+

--- a/misc/packcc/tests/ascii.d/input.txt
+++ b/misc/packcc/tests/ascii.d/input.txt
@@ -1,0 +1,1 @@
+This is a test

--- a/misc/packcc/tests/calc.d/calc.bats
+++ b/misc/packcc/tests/calc.d/calc.bats
@@ -1,14 +1,13 @@
 #!/usr/bin/env bats
 
-load $TESTDIR/utils.sh
+load "$TESTDIR/utils.sh"
 
 @test "Testing calc.d - generation" {
-    cp -f ../src/examples/calc.peg calc.d/parser.peg  ## NOTE: Copy is adopted instead of using a link, considering MinGW.
-    test_generate calc.d parser.peg
+    test_generate "$ROOTDIR/src/examples/calc.peg"
 }
 
 @test "Testing calc.d - compilation" {
-    "${CC:-cc}" calc.d/parser.c -o calc.d/parser
+    ${CC:-cc} calc.d/parser.c -o calc.d/parser
 }
 
 @test "Testing calc.d - run" {

--- a/misc/packcc/tests/code_generation.d/generation.bats
+++ b/misc/packcc/tests/code_generation.d/generation.bats
@@ -1,25 +1,13 @@
 #!/usr/bin/env bats
 
-load $TESTDIR/utils.sh
-
-in_header() {
-    grep -Fq "$1" "code_generation.d/parser.h"
-}
-
-in_source() {
-    grep -Fq "$1" "code_generation.d/parser.c"
-}
-
-get_line() {
-    sed -n "/$1/=" "code_generation.d/$2" | tail -n1
-}
+load "$TESTDIR/utils.sh"
 
 @test "Testing code_generation.d - generation" {
-    test_generate "code_generation.d"
+    test_generate
 }
 
 @test "Testing code_generation.d - compilation" {
-    test_compile "code_generation.d" \
+    test_compile \
         -D pcc_create=my_create \
         -D pcc_destroy=my_destroy \
         -D pcc_parse=my_parse \

--- a/misc/packcc/tests/dump.d/dump.bats
+++ b/misc/packcc/tests/dump.d/dump.bats
@@ -1,0 +1,8 @@
+#!/usr/bin/env bats
+
+load "$TESTDIR/utils.sh"
+
+@test "Testing dump.d - generation" {
+    run "$PACKCC" --debug -o "dump.d/parser" "$ROOTDIR/src/examples/calc.peg" 2>&1
+    check_output "dump.d/expected.txt"
+}

--- a/misc/packcc/tests/dump.d/expected.txt
+++ b/misc/packcc/tests/dump.d/expected.txt
@@ -1,0 +1,163 @@
+Rule(name:'statement', ref:0, vars.len:1, capts.len:0, codes.len:2) {
+  Alternate(max:2, len:2) {
+    Sequence(max:8, len:5) {
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'e', index:0, name:'expression', rule:'expression')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'(null)', index:void, name:'EOL', rule:'EOL')
+      Action(index:0, value:{ printf(\"answer=%d\\n\", e); }, vars:
+        'e'
+      )
+    }
+    Sequence(max:4, len:3) {
+      Quantity(min:0, max-1) {
+        Sequence(max:2, len:2) {
+          Predicate(neg:1) {
+            Reference(var:'(null)', index:void, name:'EOL', rule:'EOL')
+          }
+          Charclass(value:'null')
+        }
+      }
+      Reference(var:'(null)', index:void, name:'EOL', rule:'EOL')
+      Action(index:1, value:{ printf(\"error\\n\"); }, vars:none)
+    }
+  }
+}
+Rule(name:'expression', ref:2, vars.len:1, capts.len:0, codes.len:1) {
+  Sequence(max:2, len:2) {
+    Reference(var:'e', index:0, name:'term', rule:'term')
+    Action(index:0, value:{ __ = e; }, vars:
+      'e'
+    )
+  }
+}
+Rule(name:'term', ref:3, vars.len:3, capts.len:0, codes.len:3) {
+  Alternate(max:4, len:3) {
+    Sequence(max:8, len:6) {
+      Reference(var:'l', index:0, name:'term', rule:'term')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      String(value:'+')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'r', index:1, name:'factor', rule:'factor')
+      Action(index:0, value:{ __ = l + r; }, vars:
+        'l'
+        'r'
+      )
+    }
+    Sequence(max:8, len:6) {
+      Reference(var:'l', index:0, name:'term', rule:'term')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      String(value:'-')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'r', index:1, name:'factor', rule:'factor')
+      Action(index:1, value:{ __ = l - r; }, vars:
+        'l'
+        'r'
+      )
+    }
+    Sequence(max:2, len:2) {
+      Reference(var:'e', index:2, name:'factor', rule:'factor')
+      Action(index:2, value:{ __ = e; }, vars:
+        'e'
+      )
+    }
+  }
+}
+Rule(name:'factor', ref:5, vars.len:3, capts.len:0, codes.len:3) {
+  Alternate(max:4, len:3) {
+    Sequence(max:8, len:6) {
+      Reference(var:'l', index:0, name:'factor', rule:'factor')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      String(value:'*')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'r', index:1, name:'unary', rule:'unary')
+      Action(index:0, value:{ __ = l * r; }, vars:
+        'l'
+        'r'
+      )
+    }
+    Sequence(max:8, len:6) {
+      Reference(var:'l', index:0, name:'factor', rule:'factor')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      String(value:'/')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'r', index:1, name:'unary', rule:'unary')
+      Action(index:1, value:{ __ = l / r; }, vars:
+        'l'
+        'r'
+      )
+    }
+    Sequence(max:2, len:2) {
+      Reference(var:'e', index:2, name:'unary', rule:'unary')
+      Action(index:2, value:{ __ = e; }, vars:
+        'e'
+      )
+    }
+  }
+}
+Rule(name:'unary', ref:5, vars.len:1, capts.len:0, codes.len:3) {
+  Alternate(max:4, len:3) {
+    Sequence(max:4, len:4) {
+      String(value:'+')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'e', index:0, name:'unary', rule:'unary')
+      Action(index:0, value:{ __ = +e; }, vars:
+        'e'
+      )
+    }
+    Sequence(max:4, len:4) {
+      String(value:'-')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'e', index:0, name:'unary', rule:'unary')
+      Action(index:1, value:{ __ = -e; }, vars:
+        'e'
+      )
+    }
+    Sequence(max:2, len:2) {
+      Reference(var:'e', index:0, name:'primary', rule:'primary')
+      Action(index:2, value:{ __ = e; }, vars:
+        'e'
+      )
+    }
+  }
+}
+Rule(name:'primary', ref:1, vars.len:1, capts.len:1, codes.len:2) {
+  Alternate(max:2, len:2) {
+    Sequence(max:2, len:2) {
+      Capture(index:0) {
+        Quantity(min:1, max-1) {
+          Charclass(value:'0-9')
+        }
+      }
+      Action(index:0, value:{ __ = atoi(_1); }, vars:
+        $1
+      )
+    }
+    Sequence(max:8, len:6) {
+      String(value:'(')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      Reference(var:'e', index:0, name:'expression', rule:'expression')
+      Reference(var:'(null)', index:void, name:'_', rule:'_')
+      String(value:')')
+      Action(index:1, value:{ __ = e; }, vars:
+        'e'
+      )
+    }
+  }
+}
+Rule(name:'_', ref:14, vars.len:0, capts.len:0, codes.len:0) {
+  Quantity(min:0, max-1) {
+    Charclass(value:' \t')
+  }
+}
+Rule(name:'EOL', ref:3, vars.len:0, capts.len:0, codes.len:0) {
+  Alternate(max:4, len:4) {
+    String(value:'\n')
+    String(value:'\r\n')
+    String(value:'\r')
+    String(value:';')
+  }
+}
+value_type: 'int'
+auxil_type: 'void *'
+prefix: 'calc'

--- a/misc/packcc/tests/style.d/style.bats
+++ b/misc/packcc/tests/style.d/style.bats
@@ -10,7 +10,6 @@ check_uncrustify_version() {
 }
 
 test_style() {
-    check_uncrustify_version
     if ! command -v "uncrustify" &> /dev/null; then
         skip "uncrustify is not installed"
     elif ! check_uncrustify_version &> /dev/null; then
@@ -29,8 +28,7 @@ test_style() {
 }
 
 @test "Testing style.d - generated" {
-    cp -f ../src/examples/calc.peg style.d/parser.peg  ## NOTE: Copy is adopted instead of using a link, considering MinGW.
-    test_generate "style.d" "parser.peg"
+    test_generate "$ROOTDIR/src/examples/calc.peg"
     test_style "style.d/parser.h"
     test_style "style.d/parser.c"
 }

--- a/misc/packcc/tests/test.sh
+++ b/misc/packcc/tests/test.sh
@@ -8,14 +8,14 @@ generate_bats() {
     cat <<EOF
 #!/usr/bin/env bats
 
-load $TESTDIR/utils.sh
+load "\$TESTDIR/utils.sh"
 
 @test "Testing $1 - generation" {
-    ${skip_all}test_generate "$1"
+    ${skip_all}test_generate
 }
 
 @test "Testing $1 - compilation" {
-    ${skip_all}test_compile "$1"
+    ${skip_all}test_compile
 }
 EOF
     for input in "$1"/input*.txt; do
@@ -40,12 +40,12 @@ EOF
 build() {
     if [ -z "$PACKCC" ]; then
         export PACKCC="$TESTDIR/packcc"
-        "${CC:-cc}" -o "$PACKCC" $ROOTDIR/src/packcc.c
+        ${CC:-cc} --coverage -O0 -o "$PACKCC" $ROOTDIR/src/packcc.c
     fi
 }
 
 clean() {
-    rm -f packcc *.d/test.bats *.d/parser{,.c,.h}
+    rm -f packcc{,.gcda,.gcno,.c.gcov} *.d/test.bats *.d/parser{,.c,.h}
 }
 
 main() {
@@ -65,6 +65,10 @@ main() {
     done
 
     bats "$@" ./*.d
+
+    if [ -f "packcc.gcda" ]; then
+        echo "$(gcov packcc | grep "Lines executed") (see $TESTDIR/packcc.c.gcov for details)"
+    fi
 }
 
 main "$@"

--- a/misc/packcc/tests/utils.sh
+++ b/misc/packcc/tests/utils.sh
@@ -1,14 +1,28 @@
 test_generate () {
-    (cd "$1" && "$PACKCC" -o "parser" "${2:-input.peg}")
+    (cd "$BATS_TEST_DIRNAME" && "$PACKCC" "${PACKCC_OPTS[@]}" -o "parser" "${1:-input.peg}")
 }
 
 test_compile() {
-    local dir="$1"
-    shift
-    "${CC:-cc}" -I "$dir" "main.c" -o "$dir/parser" "$@"
+    ${CC:-cc} -I "$BATS_TEST_DIRNAME" "main.c" -o "$BATS_TEST_DIRNAME/parser" "$@"
+}
+
+check_output() {
+    diff --strip-trailing-cr -uN "${1/input/expected}" --label "${1/input/expected}" <(echo "$output") --label "output"
 }
 
 run_for_input() {
-    run timeout 5s "$(dirname "$1")/parser" < "$1"
-    diff --strip-trailing-cr -uN "${1/input/expected}" --label "${1/input/expected}" <(echo "$output") --label "output"
+    run timeout 5s "$BATS_TEST_DIRNAME/parser" < "$1"
+    check_output "$1"
+}
+
+in_header() {
+    grep -Fq "$1" "$BATS_TEST_DIRNAME/parser.h"
+}
+
+in_source() {
+    grep -Fq "$1" "$BATS_TEST_DIRNAME/parser.c"
+}
+
+get_line() {
+    sed -n "/$1/=" "$BATS_TEST_DIRNAME/$2" | tail -n1
 }

